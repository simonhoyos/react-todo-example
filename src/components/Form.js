import React from 'react';

export function Form({ title, editingId, handleChange, handleSubmit }) {
  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="title">Nombre de la tarea</label>
      <input
        type="text"
        id="title"
        name="title"
        onChange={handleChange}
        value={title}
        data-testid="input-title"
      />
      <button type="submit">{editingId > -1 ? 'Editar tarea' : 'Crear nueva tarea'}</button>
    </form>
  );
}
