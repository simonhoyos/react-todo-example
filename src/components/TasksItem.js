import React from 'react';
import { TasksConsumer } from '../Context';

export function TaskItem({ id, title, done }) {
  return (
    <TasksConsumer>
      {({ handleEdit, handleToggleTask }) => (
        <li>
          {title} - {done ? 'completa' : 'incompleta'}
          <button onClick={() => handleEdit(id, title, done)}>Editar</button>
          <button onClick={() => handleToggleTask(id)}>Completar</button>
        </li>
      )}
    </TasksConsumer>
  );
}
