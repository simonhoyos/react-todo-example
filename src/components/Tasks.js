import React from 'react';
import { TaskItem } from './TasksItem';

export function Tasks({ tasks, handleEdit, handleToggleTask }) {
  return (
    <ul>
      {tasks.length > 0 && tasks.map(({ id, title, done }) => {
        return (
          <TaskItem
            key={id}
            id={id}
            title={title}
            done={done}
            handleEdit={handleEdit}
            handleToggleTask={handleToggleTask}
          />
        );
      })}
    </ul>
  );
}
