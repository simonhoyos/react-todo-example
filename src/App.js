import React, { Component } from 'react';
import './App.css';
import { Tasks } from './components/Tasks';
import { Form } from './components/Form';

class App extends Component {
  state = {
    tasks: [
      { id: 0, title: 'una tarea', done: true },
      { id: 1, title: 'segunda tarea', done: false },
      { id: 2, title: 'tercera tarea', done: false },
      { id: 3, title: 'tercera tarea', done: false },
      { id: 4, title: 'tercera tarea', done: false },
    ],
    title: '',
    editingId: -1,
    done: false,
  };

  handleChange = e => {
    const { name, value } = e.target;

    this.setState({
      [name]: value,
    });
  }

  handleSubmit = e => {
    e.preventDefault();

    const id = this.state.editingId > -1 ?
      this.state.editingId :
      this.state.tasks.length;

    const newTask = {
      id,
      title: this.state.title,
      done: this.state.done,
    };

    let tasks;
    if (this.state.editingId < 0) {
      tasks = [ ...this.state.tasks, newTask ];
    } else {
      tasks = this.state.tasks.map((task) => {
        if (task.id === id) {
          return {
            ...task,
            ...newTask,
          }
        }

        return task;
      });
    }

    this.setState({
      tasks,
      title: '',
      editingId: -1,
      done: false,
    });
  }

  handleEdit = (id, title, done) => {
    this.setState({
      editingId: id,
      title,
      done
    });
  }

  handleToggleTask = id => {
    const tasks = this.state.tasks.map(task => {
      if (task.id === id) {
        return {
          ...task,
          done: !task.done,
        }
      }

      return task;
    });

    this.setState({ tasks });
  }

  render() {
    const { tasks, title, editingId } = this.state;
    return (
      <div className="App">
        <Tasks
          tasks={tasks}
          handleEdit={this.handleEdit}
          handleToggleTask={this.handleToggleTask}
        />
        <Form
          title={title}
          editingId={editingId}
          handleChange={this.handleChange}
          handleSubmit={this.handleSubmit}
        />
      </div>
    );
  }
}

export default App;
