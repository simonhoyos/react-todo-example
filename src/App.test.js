import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { TaskItem } from './components/TasksItem';

describe('<App />', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<App />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should edit a existing task', () => {
    const wrapper = mount(<App />);
    const item = wrapper.find('li').first();
    const itemContent = item.prop('children')[0]; // una tarea
    item.find('[data-testid="edit-button"]').simulate('click');

    const input = wrapper.find('[data-testid="input-title"]');

    expect(input.prop('value')).toBe(itemContent);

    const newValue = 'tarea editada';
    input.simulate('change', {
      target: {
        name: 'title',
        value: newValue,
      }
    });

    const form = wrapper.find('form');
    form.simulate('submit');

    const newItem = wrapper.find('li').first();
    expect(newItem.prop('children')[0]).toBe(newValue);
  });

  it('should create a new task', () => {
    const wrapper = mount(<App />);
    const input = wrapper.find('[data-testid="input-title"]');
    const newValue = 'tarea editada';
    const intialTasksLength = wrapper.state().tasks.length;

    input.simulate('change', {
      target: {
        name: 'title',
        value: newValue,
      }
    });

    const form = wrapper.find('form');
    form.simulate('submit');

    expect(wrapper.state().tasks).toHaveLength(intialTasksLength + 1);
    expect(wrapper.state().tasks[intialTasksLength].title).toBe(newValue);

    wrapper.update();

    expect(wrapper.find(TaskItem)).toHaveLength(intialTasksLength + 1);
  });

  it('should change task status', () => {
    const wrapper = mount(<App />);
    const button = wrapper.find('[data-testid="toggle-button"]').first();

    button.simulate('click');

    const item = wrapper.find('li').first();

    expect(item.props().children).toContain('incompleta');
  });


  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
});
