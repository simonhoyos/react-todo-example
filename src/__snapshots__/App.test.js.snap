// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match snapshot 1`] = `
<div
  className="App"
>
  <Tasks
    handleEdit={[Function]}
    handleToggleTask={[Function]}
    tasks={
      Array [
        Object {
          "done": true,
          "id": 0,
          "title": "una tarea",
        },
        Object {
          "done": false,
          "id": 1,
          "title": "segunda tarea",
        },
        Object {
          "done": false,
          "id": 2,
          "title": "tercera tarea",
        },
        Object {
          "done": false,
          "id": 3,
          "title": "tercera tarea",
        },
        Object {
          "done": false,
          "id": 4,
          "title": "tercera tarea",
        },
      ]
    }
  />
  <Form
    editingId={-1}
    handleChange={[Function]}
    handleSubmit={[Function]}
    title=""
  />
</div>
`;
